File: chat-service/src/main/java/com/relay/chatservice/user/UserController.java
package com.relay.chatservice.user;

import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.messaging.handler.annotation.MessageMapping;
import org.springframework.messaging.handler.annotation.Payload;
import org.springframework.messaging.handler.annotation.SendTo;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.GetMapping;

import java.util.List;

@Service
@RequiredArgsConstructor
public class UserController {
    private final UserService userService;

    @MessageMapping("/user.addUser")
    @SendTo("/user/topic")
    public User addUser(
            @Payload User user
    ) {
        userService.saveUser(user);
        return user;
    }

    @MessageMapping("/user.disconnect")
    @SendTo("/user/topic")
    public User disconnect(
            @Payload User user
    ) {
        userService.disconnect(user);
        return user;
    }

    @GetMapping("/users")
    public ResponseEntity<List<User>> findConnectedUsers() {
        return ResponseEntity.ok(userService.findConnectedUsers());
    }
}

File: chat-service/src/main/java/com/relay/chatservice/user/UserController.kt
package com.relay.chatservice.user

import lombok.RequiredArgsConstructor
import org.springframework.http.ResponseEntity
import org.springframework.messaging.handler.annotation.MessageMapping
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.messaging.handler.annotation.SendTo
import org.springframework.stereotype.Service
import org.springframework.web.bind.annotation.GetMapping

@Service
@RequiredArgsConstructor
class UserController(private val userService: UserService) {

    @MessageMapping("/user.addUser")
    @SendTo("/user/topic")
    fun addUser(
        @Payload user: User
    ): User {
        userService.saveUser(user)
        return user
    }

    @MessageMapping("/user.disconnect")
    @SendTo("/user/topic")
    fun disconnect(
        @Payload user: User
    ): User {
        userService.disconnect(user)
        return user
    }

    @GetMapping("/users")
    fun findConnectedUsers(): ResponseEntity<MutableList<User?>?> {
        return ResponseEntity.ok<MutableList<User?>?>(userService.findConnectedUsers())
    }
}
